### PR TITLE
docs(template-syntax/Dart): TrackBy code + fixes.

### PR DESCRIPTION
**Pull request by thso**
Tried to stay as close as possible to the ts version, while showcasing nice features of Dart.

  - Implementation of elements change counters used for the comparison of no `TrackBy` vs proper usage of `TrackBy`,
  - Fixed issues that prevented from compiling, 
  - General style cleanup.

Can you take look @kwalrath ?

### CODE REVIEW

File: public/docs/_examples/template-syntax/dart/lib/hero_detail_component.dart [hunk 1]
```diff
@@ -56,18 +57,17 @@ class HeroDetailComponent {
...
+        <div>Last: {{hero?.lastName}}</div>
+        <div>Birthdate: {{hero?.birthdate | date:'longDate'}}</div>
+        <div>Web: <a href="{{hero?.url}}" target="_blank">{{hero?.url}}</a></div>
+        <div>Rate/hr: {{hero?.rate | currency:'EUR'}}</div>
+        <br clear="all">
+        <button (click)="delete()">Delete</button>
+      </div>''')
 class BigHeroDetailComponent extends HeroDetailComponent {
   // #docregion input-output-1
   @Input() Hero hero;
```
Comment (kwalrath) [inline public/docs/_examples/template-syntax/dart/lib/hero_detail_component.dart]: Do you think we should have `@override` here?

**Comment by kwalrath** [comment]
We have some no-return methods that aren't marked void. Should we say their return type is void? I'm thinking of delete() in [Big]HeroDetailComponent, and ng*Init() in AppComponent.

**Comment by kwalrath** [comment]
We'd talked about changing the `ngSwitchWhen` templates (in app_component.html) to not have the `[...]`. For example, we'd change this:

```
<template [ngSwitchWhen]="'Eenie'"><span>Eenie</span></template>
```
to this:
```
<template ngSwitchWhen="Eenie"><span>Eenie</span></template>
```

The former works, but it's ugly.

**Comment by thso** [comment]
> We have some no-return methods that aren't marked void. Should we say their return type is void? I'm thinking of delete() in [Big]HeroDetailComponent, and ng*Init() in AppComponent.

Yes, we should add return types to every function definition as it default to dynamic, which is bad.

File: public/docs/_examples/template-syntax/dart/lib/hero_detail_component.dart (hunk 1 above)
Comment (thso) [inline public/docs/_examples/template-syntax/dart/lib/hero_detail_component.dart]: Yup, I'm on it.

**Comment by thso** [comment]
I agree `<template ngSwitchWhen="Eenie"><span>Eenie</span></template>` is cleaner.

**Comment by wardbell** [comment]
If hand coded it would be: `<span ngSwitchWhen="Eenie">Eenie</span>`

**Comment by wardbell** [comment]
The key is that one should use `foo="some-string"` when you're NOT binding and reserve `[foo]="someProperty"` for property binding scenarios in which the value must be watched.

**Comment by kwalrath** [comment]
@thso: Yes, please change it to `<template ngSwitchWhen="Eenie"><span>Eenie</span></template>` is the upshot. (Right, @wardbell?)

File: public/docs/_examples/template-syntax/dart/lib/app_component.dart [hunk 2]
```diff
@@ -19,7 +19,17 @@ enum Color { Red, Green, Blue }
       MyClickDirective,
       MyClickDirective2
     ])
-class AppComponent {
+class AppComponent implements OnInit, AfterViewInit {
+  @override
+  ngOnInit() {
```
Comment (zoechi) [inline public/docs/_examples/template-syntax/dart/lib/app_component.dart]: The lifecycle methods don't have return types (here and below) is this intentional?

**Comment by wardbell** [comment]
Well .. that's what I did and why. I would think the same reasoning would be persuasive in Dart world but not sure. That's your judgement call

**Comment by kwalrath** [comment]
LGTM. Thanks, @thso!

**Pull request closed without merging**